### DISCUSSION TITLE
manifest.json not found

### DISCUSSION THREAD

**Discussion by Sa1ch** [adonisjs/core#4712] [Help] [answered]
Hi! I'm moving the application to the server, but I'm getting an error ENOENT: no such file or directory, open 'public/assets/.vite/manifest.json'.
The file is present, and if I call it directly, for example: https://example.com/assets/.vite/manifest.json, it opens successfully.
In home.edge, there is the following line: @vite(['resources/js/app.js']).
In vite.config.ts: adonisjs({ entrypoints: ['resources/js/app.js'], reload: ['resources/views/**/*.edge'] }).

I don't understand what else this error could be related to. Has anyone encountered a similar situation?

**Comment by RomainLanz** [comment]
Hey! 👋🏻 

Are you trying to deploy to production?
Can you ensure that you are running your code with `NODE_ENV=production`?

**Reply by Sa1ch** [reply]
Yes, in my .env file, I have NODE_ENV=production set.
I copy the build folder to the server from my local computer and run npm install.

Am I doing everything correctly? Or do I need to copy the entire project?

The thing is, I haven't worked with Node.js before, but I have worked with PHP (Laravel), so I might not understand something.

**Comment by Sa1ch** [comment]
I cloned the entire project to the server and ran the build. But it didn't help; the same error occurred.

**Comment by Sa1ch** [comment, marked as answer]
The following helped me:

npx npm-check-updates -u
npm install

**Comment by eduwass** [comment]
I'm encountering the same issue, I ran the commands above but that didn't help.

The file seems to be in place and responds if accessed directly:
```
root@fe2352e8756a:/app/build/public/assets/.vite# pwd
/app/build/public/assets/.vite
root@fe2352e8756a:/app/build/public/assets/.vite# ls -lah
total 12K
drwxr-xr-x 2 root root 4.0K Sep 18 08:35 .
drwxr-xr-x 3 root root 4.0K Sep 18 08:35 ..
-rw-r--r-- 1 root root 1.2K Sep 18 08:35 manifest.json

root@fe2352e8756a:/app/build/public/assets/.vite# curl http://localhost:3333/assets/.vite/manifest.json -I
HTTP/1.1 200 OK
Accept-Ranges: bytes
Cache-Control: public, max-age=0
Last-Modified: Wed, 18 Sep 2024 08:35:11 GMT
ETag: W/"489-19204456d4f"
Content-Type: application/json; charset=UTF-8
Content-Length: 1161
Date: Wed, 18 Sep 2024 08:39:32 GMT
Connection: keep-alive
Keep-Alive: timeout=5
```

Also NODE_ENV is set to Production.


The project was started from the API starter kit, but I later decided to add Vite, Inertia, Edge for the frontend, seems to all work well in my local but showing this error when deployed.

Any ideas @RomainLanz @Sa1ch ?

Thanks!

**Reply by Sa1ch** [reply]
Hello! Try running the command `npm audit`
And if any issues are found, run `npm audit fix`
This also helped me: https://adonisjs.com/blog/upgrading-eslint-typescript-setup.

**Comment by eduwass** [comment]
I had to update the Vite configuration on file `config/vite.ts` like so:

```ts
import { defineConfig } from '@adonisjs/vite'
import app from '@adonisjs/core/services/app'

const viteBackendConfig = defineConfig({
  /**
   * The output of vite will be written inside this
   * directory. The path should be relative from
   * the application root.
   */
  buildDirectory: 'public/assets',

  /**
   * The path to the manifest file generated by the
   * "vite build" command.
   */
  manifestFile: `${app.inProduction ? 'build/' : ''}public/assets/.vite/manifest.json`,

  /**
   * Feel free to change the value of the "assetsUrl" to
   * point to a CDN in production.
   */
  assetsUrl: '/assets',

  scriptAttributes: {
    defer: true,
  },
})

export default viteBackendConfig
```

The key is to tell the `manifestFile` path to include **`build/`** prefix when running in production mode.

**Reply by Sa1ch** [reply]
I found another reason why this might be happening. I encountered this error again, and it turned out to be an issue with the configuration in the ecosystem.config.js file for pm2.

**Reply by GradeyCullins** [reply]
> I had to update the Vite configuration on file `config/vite.ts` like so:
> 
> ```ts
> import { defineConfig } from '@adonisjs/vite'
> import app from '@adonisjs/core/services/app'
> 
> const viteBackendConfig = defineConfig({
>   /**
>    * The output of vite will be written inside this
>    * directory. The path should be relative from
>    * the application root.
>    */
>   buildDirectory: 'public/assets',
> 
>   /**
>    * The path to the manifest file generated by the
>    * "vite build" command.
>    */
>   manifestFile: `${app.inProduction ? 'build/' : ''}public/assets/.vite/manifest.json`,
> 
>   /**
>    * Feel free to change the value of the "assetsUrl" to
>    * point to a CDN in production.
>    */
>   assetsUrl: '/assets',
> 
>   scriptAttributes: {
>     defer: true,
>   },
> })
> 
> export default viteBackendConfig
> ```
> 
> The key is to tell the `manifestFile` path to include **`build/`** prefix when running in production mode.

This worked for me. Much thanks!

**Comment by atakansn** [comment]
I've tried both methods but it doesn't work @GradeyCullins @Sa1ch  @RomainLanz

**Reply by RomainLanz** [reply]
Ensure that you are running your code in production mode using `NODE_ENV=production`.

**Comment by ericpugh** [comment]
Just in case it helps someone, I had the same issue caused by a problem with my PM2 config file. Basically, PM2 was running from the wrong root directory with the following config (which, at the time, was being started from a shared directory outside the code base:

```js
module.exports = {
  apps: [
    {
      name: 'nodejs-app',
      script: '/var/www/my.domain.com/current/build/bin/server.js',
      instances: 'max',
      exec_mode: 'cluster',
      autorestart: true,
      env: {
      'NODE_ENV': 'production',
     },
    },
  ],
}
```

I changed it to this to get it working. (Notice the "cwd" property)


```js
module.exports = {
  apps: [
    {
      name: 'nodejs-app',
     cwd: '/var/www/my.domain.com/current/build',
      script: 'bin/server.js',
      instances: 'max',
      exec_mode: 'cluster',
      autorestart: true,
      env: {
      'NODE_ENV': 'production',
     },
    },
  ],
}
```
